### PR TITLE
Unwrap DDP model before loss calculation

### LINKED ISSUE #22015: problem to learn model at 2 GPU's | device=[0,1]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, in the new version of ultralytics, I have encountered the problem of training on two GPUs. There is no such problem in the old version. Can you tell me if there are new hyperparameters or a new way to parallelize the process?

Image

### Additional

_No response_

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #22015 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Ensures loss is computed on the underlying model by unwrapping any wrappers, improving compatibility with torch.compile and other training wrappers. ✅

### 📊 Key Changes
- Replaced `self.model.loss(...)` with `unwrap_model(self.model).loss(...)` during training loss computation.

### 🎯 Purpose & Impact
- Improves compatibility with torch.compile and other wrappers (e.g., DDP/EMA) that may hide or alter `.loss` on the wrapped model. 🧩
- Reduces runtime errors and edge cases when using compiled or wrapped models, leading to more reliable training. 🚀
- No expected change to training results or user-facing APIs—behavior remains the same, just more robust. 🔒

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response — an Ultralytics engineer will review and assist soon. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

PR-specific notes for “Unwrap DDP model before loss calculation”:
- 🔎 Please confirm that the change maintains identical logging and `self.loss_items` behavior across single-GPU, DDP, and EMA scenarios.
- 🧪 If possible, add or update tests that exercise compiled training paths and DDP-wrapped models to prevent regressions.
- 🧰 Ensure imports and references are consistent throughout the training loop and compatible with torch.compile, and that all CI jobs pass.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by Y-T-G** [comment]
@glenn-jocher @Laughing-q I wonder if it's better to just create and use a custom wrapper that fallbacks to method of the wrapped model, like how we do with `Model` class:

```python
class DistributedDataParallel(torch.nn.parallel.DistributedDataParallel):
    def __getattr__(self, name):
        try:
            return super().__getattr__(name)
        except AttributeError:
            return getattr(self.module, name)
```

**Comment by Laughing-q** [comment]
@Y-T-G wait should we unwrap the model or the model.loss? I saw you are unwrapping model.loss

**Comment by Laughing-q** [comment]
> @glenn-jocher @Laughing-q I wonder if it's better to just create and use a custom wrapper that fallbacks to method of the wrapped model, like how we do with `Model` class:
> 
> ```python
> class DistributedDataParallel(torch.nn.parallel.DistributedDataParallel):
>     def __getattr__(self, name):
>         try:
>             return super().__getattr__(name)
>         except AttributeError:
>             return getattr(self.module, name)
> ```

yeah this sounds interesting, can you have a PR for it?

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
@glenn-jocher need ddp ci tests back asap.

**Comment by UltralyticsAssistant** [comment]
🎉 Merged! Huge thanks to @Y-T-G and @Laughing-q for the thoughtful improvement in [PR #22016](https://github.com/ultralytics/ultralytics/pull/22016).

“Simplicity is the ultimate sophistication.” — Leonardo da Vinci  
By unwrapping the model for loss computation, you’ve made training with torch.compile and wrappers like DDP/EMA more reliable—keeping behavior unchanged while strengthening the core. Subtle change, big impact. 🙌

Appreciate the attention to detail and commitment to robustness. This makes our training pipeline smoother for everyone across the Ultralytics community. 🚀
